oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoDocumentStoreThrottling.java

```diff
@@ -0,0 +1,99 @@
+    static final String TS_TIME = "ts";
+    private static final String NATURAL = "$natural";
+    private static final String MAX_SIZE = "maxSize";
+    private static final String OPLOG_RS = "oplog.rs";
+    public static final String SIZE = "size";
+    private final ScheduledExecutorService throttlingExecutor;
+    private final AtomicDouble oplogWindow;
+    private final MongoDatabase localDb;
+
+    public MongoDocumentStoreThrottling(final @NotNull MongoDatabase localDb, final @NotNull MongoThrottlingMetrics mongoThrottlingMetrics) {
+        this.throttlingExecutor = Executors.newSingleThreadScheduledExecutor();
+        this.oplogWindow = mongoThrottlingMetrics.oplogWindow;
+        this.localDb = localDb;
+    }
+
+    public void updateMetrics() {
+        throttlingExecutor.scheduleAtFixedRate(() -> {
+            Document document = localDb.runCommand(new Document("collStats", OPLOG_RS));
+            if (!document.containsKey(MAX_SIZE) || !document.containsKey(SIZE)) {
+                LOG.warn("Could not get stats for local.{}  collection. collstats returned: {}.", OPLOG_RS, document);
+            } else {
+                int maxSize = document.getInteger(MAX_SIZE);
+                double maxSizeGb = (double) maxSize / (1024 * 1024 * 1024);
+                int usedSize = document.getInteger(SIZE);
+                double usedSizeGb = Math.ceil(((double) usedSize / (1024 * 1024 * 1024)) * 1000000) / 1000000;
+                MongoCollection<Document> localDbCollection = localDb.getCollection(OPLOG_RS);
+                Document first = localDbCollection.find().sort(new Document(NATURAL, 1)).limit(1).first();
+                Document last = localDbCollection.find().sort(new Document(NATURAL, -1)).limit(1).first();
+
+                if (Objects.isNull(first) || Objects.isNull(last)) {
+                    LOG.warn("Objects not found in local.oplog.rs -- is this a new and empty db instance?");
+                } else {
+                    if (!first.containsKey(TS_TIME) || !last.containsKey(TS_TIME)) {
+                        LOG.warn("ts element not found in oplog objects");
+                    } else {
+                        oplogWindow.set(updateOplogWindow(maxSizeGb, usedSizeGb, first, last));
+                    }
+                }
+            }
+        }, 10, 30, SECONDS);
+    }
+
+    // helper methods
+    @VisibleForTesting
+    double updateOplogWindow(final double maxSize, final double usedSize, final @NotNull Document first,
+                                   final @NotNull Document last) {
+        final BsonTimestamp startTime = first.get(TS_TIME, BsonTimestamp.class);
+        final BsonTimestamp lastTime = last.get(TS_TIME, BsonTimestamp.class);
+        long timeDiffSec = Math.abs(lastTime.getTime() - startTime.getTime());
+        double timeDiffHr = Math.ceil(((double)timeDiffSec/(60*60)) * 100000)/100000;
+        double currentOplogHourRate = usedSize /timeDiffHr;
```

Comment:
Actually, I'm wrong. This won't throw an exception when `timeDiffSec` and `timeDiffHr` are zero, because `timeDiffHr` is of type `double`. In that case `currentOplogHourRate` will be `Infinity` and `timeLeft` zero.